sway/config/output.c

```diff
@@ -513,125 +504,43 @@ static void default_output_config(struct output_config *oc,
-	if (config->reloading) {
-		default_output_config(result, sway_output->wlr_output);
-	}
-	if (oc_id_on_name) {
-		// Already have an identifier on name config, use that
-		free(result->name);
-		result->name = strdup(id_on_name);
-		merge_output_config(result, oc_id_on_name);
-	} else if (oc_name && oc_id) {
-		// Generate a config named `<identifier> on <name>` which contains a
-		// merged copy of the identifier on name. This will make sure that both
-		// identifier and name configs are respected, with identifier getting
-		// priority
-		struct output_config *temp = new_output_config(id_on_name);
-		merge_output_config(temp, oc_name);
-		merge_output_config(temp, oc_id);
-		list_add(config->output_configs, temp);
-
-		free(result->name);
-		result->name = strdup(id_on_name);
-		merge_output_config(result, temp);
-
-		sway_log(SWAY_DEBUG, "Generated output config \"%s\" (enabled: %d)"
-			" (%dx%d@%fHz position %d,%d scale %f transform %d) (bg %s %s)"
-			" (dpms %d) (max render time: %d)", result->name, result->enabled,
-			result->width, result->height, result->refresh_rate,
-			result->x, result->y, result->scale, result->transform,
-			result->background, result->background_option, result->dpms_state,
-			result->max_render_time);
-	} else if (oc_name) {
-		// No identifier config, just return a copy of the name config
-		free(result->name);
-		result->name = strdup(name);
-		merge_output_config(result, oc_name);
-	} else if (oc_id) {
-		// No name config, just return a copy of the identifier config
-		free(result->name);
-		result->name = strdup(identifier);
-		merge_output_config(result, oc_id);
-	} else {
-		i = list_seq_find(config->output_configs, output_name_cmp, "*");
-		if (i >= 0) {
-			// No name or identifier config, but there is a wildcard config
-			free(result->name);
-			result->name = strdup("*");
-			merge_output_config(result, config->output_configs->items[i]);
-		} else if (!config->reloading) {
-			// No name, identifier, or wildcard config. Since we are not
-			// reloading with defaults, the output config will be empty, so
-			// just return NULL
-			free_output_config(result);
-			result = NULL;
-		}
-	}
+	struct output_config *oc = result;
+	sway_log(SWAY_DEBUG, "Found output config %s (enabled: %d) (%dx%d@%fHz "
+		"position %d,%d scale %f subpixel %s transform %d) (bg %s %s) (dpms %d) "
+		"(max render time: %d)",
+		oc->name, oc->enabled, oc->width, oc->height, oc->refresh_rate,
+		oc->x, oc->y, oc->scale, sway_wl_output_subpixel_to_string(oc->subpixel),
+		oc->transform, oc->background, oc->background_option, oc->dpms_state,
+		oc->max_render_time);
 
-	free(id_on_name);
 	return result;
 }
 
-struct output_config *find_output_config(struct sway_output *output) {
-	char id[128];
-	output_get_identifier(id, sizeof(id), output);
-	return get_output_config(id, output);
-}
-
-void apply_output_config_to_outputs(struct output_config *oc) {
+void apply_output_config_to_outputs(void) {
```

Comment:
If this is a problem, it would be simple to add a `struct output_config` to each output that stores whatever is the current output configuration so that if the new one is the same the change is turned into a no-op.

What would be the worry? Flickering if the underlying wlroots code does something with the hardware even though no config has been changed?

Reply:
wlroots seems to already turn into no-ops changes that don't change anything so flickering shouldn't be possible:

https://github.com/swaywm/wlroots/blob/28cedb56237efdf388fed713ddb401a5d8d07434/types/wlr_output.c#L164-L166

Doing this at the wlroots level seems reasonable too. Instead of trying to have sway keep more state to try and avoid issuing unneeded commands.

Reply:
Yes, re-applying the same configuration should be perfectly fine. wlroots handles this so that compositors don't need to worry about it.